### ISSUE TITLE
have addon/setting for using markdownlint

### ISSUE THREAD

**Issue by chrisbeardy** [marktext/marktext#2273]
### Describe your feature request

It would be great if marktext could integrate with [markdownlint](https://github.com/DavidAnson/markdownlint). Either bundled with the program as a simple on / off setting, or as a downloadable addon.

I think it would be great to help allow people to get consistency with their markdown. I currently use the linter as a [vscode extension] (https://github.com/DavidAnson/vscode-markdownlint) and find it quite useful.

**Comment by brainchild0** [comment]
I think that the request is part of the larger question of what kinds of modification Mark Text makes to existing Markdown text.  Currently, some of the operation of the application I would characterize loosely as  automatic  linting, which some users dislike (myself included).  However, linting generally involves showing the user style or syntax problems in a source document, whereas Mark Text seeks to show a  presentation of  an abstract document, which would be serialized to correct Markdown source on a file save operation.


Did you have in mind a   process triggered manually or automatically?  What would you request to occur when the operation is invoked?


Linting support is good to discuss, but needs to be considered in the broader context, which includes ##2189.

**Comment by chrisbeardy** [comment]
Personally I would prefer my editor to not make changes on my behalf without my express consent. I am not a fan of the automatic changes in ordered lists for example.

For me marktext should present the document without changing it, then like any other editor, the raw underlying markdown source file should have the text added to it as typed in by the user (which marktext then displays),  it then has all the excellent tools to add markdown elements to the source file via shortcuts, menu etc (in which case it chooses the format). Unless explicitly requested by the user (option has been turned on or a format button as you suggest in #2189 ), I don't think marktext should edit the text already in the source file or overide what the user has typed in (like it does with unordered lists).

This is what I would propose linting to be for, not an automatic conversion process, but as an addon or option that if turned on, displays a list of errors in the markdown, more akin to a IDE which the user can choose to address or ignore as they desire. Clicking a format button would automatically address such issues.

**Comment by brainchild0** [comment]
Displaying a "list of errors in the Markdown" is hard to visualize in the  main application view, which shows a formatted representation of the document, not the Markdown source. 

A feature that decorates or annotates the actual Markdown would probably be an enhancement specifically to the source code editor.  Does this idea approach what you have in mind?

Mark Text is different from an IDE in that the former does not emphasize display of code to the user, but source code editing currently has limited support, and it is useful to consider directions for growth.

**Comment by chrisbeardy** [comment]
I agree, having a display of errors would clutter the main view, and as you say marktext is not an IDE. 

I like the idea of decorating the output, e.g. if there's an error put a yellow wavy underline and then maybe a tooltip which displays the error.

**Comment by brainchild0** [comment]
I'm not sure whether I am following. In "decorate the output", does "output" mean the source? Like I say, I can imagine how your basic idea might work when the source code is showing, but otherwise I find it unclear how the parts would align.

**Comment by chrisbeardy** [comment]
sorry, I mean decorate the markdown/rendered view, so the user would see something like ![image](https://user-images.githubusercontent.com/20585410/86524378-e7eef300-be71-11ea-9db4-f63e8c09f297.png) when using marktext if they were to enter something that broke the linting rules.

**Comment by chrisbeardy** [comment]
I think it should display when the source code is showing or not (should the user want linting on). I guess that would be open to preference.

**Comment by brainchild0** [comment]
Looking at the `markdownlint` rules, I find that most, as I would expect, pertain to white space, such as indenting and trailing characters. The causes of many of these problems are outside user control in regular editing. For example, when the user types *hyphen*, then *space*, Mark Text immediately formats a bullet list, but creates the indention following internal rules. In such a case, any problem exposed by the linter would represent a problem in the application, but would have no value to the user. With text created externally, then opened in Mark Text, these kinds of problems are more likely, but then the useful tool would seem to be one that simply rewrites the Markdown following internal rules (which Mark Text already does automatically, as discussed). 

I think a good starting point might be documenting a simple sequence of input in the editing buffer that would create a problem that you would want exposed by the proposed feature, and that is not presently handled satisfactorily.

**Comment by Hex5DA** [comment]
hey, is there any update on markdown lint support? i would find this very useful.

**Comment by vimkim** [comment]
This app is amazing and I'm also looking forward to this feature.
